bids/reports/report.py

```diff
@@ -0,0 +1,115 @@
+"""Generate publication-quality data acquisition methods section from BIDS dataset.
+"""
+from __future__ import print_function
+import json
+from os.path import dirname
+from os.path import abspath
+from os.path import join as pathjoin
+from collections import Counter
+
+from bids.reports import utils
+
+
+class BIDSReport(object):
+    """
+    Generates publication-quality data acquisition methods section from BIDS
+    dataset.
+
+    Parameters
+    ----------
+    layout : :obj:`bids.grabbids.BIDSLayout`
+        Layout object for a BIDS dataset.
+    config : :obj:`str` or :obj:`dict`, optional
+        Configuration info for methods generation. Can be a path to a file
+        (str), a dictionary, or None. If None, loads and uses default
+        configuration information.
+    """
+    def __init__(self, layout, config=None):
+        self.layout = layout
+        if config is None:
+            root = dirname(abspath(__file__))
+            config = pathjoin(root, 'config', 'converters.json')
+            with open(config) as fobj:
+                config = json.load(fobj)
+
+        if not isinstance(config, dict):
+            raise ValueError('Input config must be None, dict, or path to json '
+                             'file containing dict.')
+
+        self.config = config
+
+    def generate(self, task_converter=None):
+        """Generate the methods section.
+
+        Parameters
+        ----------
+        task_converter : :obj:`dict`
+            A dictionary with information for converting task names from BIDS
+            filename format to human readable strings.
+        """
+        if task_converter is None:
+            task_converter = {'rest': 'resting state'}
+
+        self.config['task'] = task_converter
+
+        descriptions = []
+
+        subjs = self.layout.get_subjects()
+        for sid in subjs:
+            description = self._report(subj=sid)
+            descriptions.append(description)
+        counter = Counter(descriptions)
+        print('Number of patterns detected: {0}'.format(len(counter.keys())))
+        print(utils.warnings())
+        return counter
+
+    def _report(self, subj):
+        """Write a report.
+
+        Parameters
+        ----------
+        subj : :obj:`str`
+            Subject ID.
+
+        Attributes
+        ----------
+        layout : :obj:`bids.grabbids.BIDSLayout`
+            Layout object for a BIDS dataset.
+        config : :obj:`dict`
+            Configuration info for methods generation.
+
+        Returns
+        -------
+        description : :obj:`str`
+            A publication-ready report of the dataset's data acquisition
+            information. Each scan type is given its own paragraph.
+        """
+        description_list = []
+        sessions = self.layout.get_sessions()
+        if sessions:
```

Comment:
I think you can simplify the next ~10 lines by requesting the niftis without the `session` constraint and asking for `return_type='obj'` or `return_type='tuple'`. Then you can inspect the returned object to see if it has a `.session` set (or, for `'obj'`, it will have a key in `.entities`), and add anything you need (and from the looks of it, you could potentially just pass it through to `utils.parse_niftis` even if it's `None`).

Reply:
I'll try doing this, but would it be possible for `BIDSLayout.get` to treat keyword arguments with `None` as if they weren't included? That would make all of this a lot easier. I'm not familiar with `grabbit`, though, so I have no clue if it's feasible.

Reply:
No, the `get` call can't ignore args with value `None`, but you can probably consolidate all the arguments in kwargs before passing, and then do something like `kwargs = {k: v for k, v in kwargs.items() if v is not None}